dist/tools/benchmark_udp/main.c

```diff
@@ -0,0 +1,273 @@
+static bool one_way;
+static uint32_t cookie;
+static uint32_t delay_us    = 100 * US_PER_MS; /* 100 ms */
+static uint16_t payload_len = 32;
+
+static char addr_str[INET6_ADDRSTRLEN];
+
+static bench_client_t *_find_or_add(list_node_t *head, struct sockaddr_in6 *addr,
+                                    bool *new_node)
+{
+    for (list_node_t* n = head->next; n; n = n->next) {
+        bench_client_t *node = container_of(n, bench_client_t, node);
+
+        if (memcmp(&addr->sin6_addr, &node->addr.sin6_addr, sizeof(addr->sin6_addr)) == 0) {
+            *new_node = false;
+            return node;
+        }
+    }
+
+    inet_ntop(AF_INET6, &addr->sin6_addr, addr_str, INET6_ADDRSTRLEN);
+    printf("adding [%s]:%d\n", addr_str, ntohs(addr->sin6_port));
+    bench_client_t *node = calloc(1, sizeof(bench_client_t));
+    memcpy(&node->addr, addr, sizeof(*addr));
+    list_add(head, &node->node);
+
+    *new_node = true;
+    return node;
+}
+
+static void clrscr(void)
+{
+    printf("\e[1;1H\e[2J");
+}
+
+static uint64_t _tv_diff_msec(struct timeval *a, struct timeval *b)
+{
+    return (a->tv_sec - b->tv_sec) * MS_PER_SEC
+         + (a->tv_usec - b->tv_usec) / US_PER_MS;
+}
+
+static void _print_stats(list_node_t *head, struct timeval *now)
+{
+    static uint8_t max_addr_len;
+
+    printf("host%*s\tbandwith\tnum TX\tnum RX",  max_addr_len - 4, "");
+    if (!one_way) {
+        printf("\t\tnum RT\t\tRTT");
+    }
+    puts("");
+
+    for (list_node_t* n = head->next; n; n = n->next) {
+        bench_client_t *node = container_of(n, bench_client_t, node);
+        uint8_t addr_len;
+
+        inet_ntop(AF_INET6, &node->addr.sin6_addr, addr_str, INET6_ADDRSTRLEN);
+        addr_len = printf("%s", addr_str);
+
+        if (addr_len > max_addr_len) {
+            max_addr_len = addr_len;
+        }
+
+        unsigned bw = (node->count_rx * (payload_len + sizeof(benchmark_msg_ping_t)))
+                    / (1 + now->tv_sec - node->first_seen.tv_sec);
+
+        unsigned success_rate = node->count_tx
+                              ? (100 * node->count_rx) / node->count_tx
+                              : 0;
+
+        printf("%*s\t%4u b/s\t%u\t%u (%u%%)",
+                                max_addr_len - addr_len, "",
+                                bw,
+                                node->count_tx, node->count_rx,
+                                success_rate);
+        if (!one_way) {
+            unsigned success_rate_rt = node->count_tx
+                                     ? (100 * node->count_rt) / node->count_tx
+                                     : 0;
+            printf("\t%u (%u%%)\t%u µs", node->count_rt, success_rate_rt, node->rtt_us);
+        }
+        puts("");
+    }
+}
+
+static void dispatch_loop(int sock)
+{
+    list_node_t head = { .next = NULL };
+    struct timeval tv_now, tv_last = { 0 };
+
+    puts("entering loop…");
+    while (1) {
+        uint8_t buffer[BENCH_PAYLOAD_SIZE_MAX + sizeof(benchmark_msg_ping_t)];
```

Comment:
?
how about allocating it using calloc according do the cmdline-parameters (make this al little more flexible)
?

Reply:
hm but then I won't be able to receive messages from nodes that were previously configured with a larger payload

Reply:
if i read the manpage correctly recv truncates the message -> only ping head will get through  -> new node -> new config

Reply:
man udp(7) 
```
 When the packet is smaller than the passed buffer, only that much data is returned; when it is bigger, the packet is truncated and the MSG_TRUNC flag is set.

```